src/team_lib/mhba/xccl_mhba_socket_comm.c

```diff
@@ -0,0 +1,277 @@
+{
+    int sock, i, num_of_curr_connections = 0;
+
+    struct sockaddr_storage storage= {};
+    struct sockaddr_un *addr;
+
+    connection_t connection[num_of_connections];
+    pthread_t thread[num_of_connections];
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1)
+    {
+        xccl_mhba_error("Failed to create server socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&storage;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, SOCKPATH, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (bind(sock, (struct sockaddr *)addr, sizeof(struct sockaddr_un)) == -1) {
+        xccl_mhba_error("Failed to bind server socket errno %d", errno);
+        goto fail;
+    }
+    if (listen(sock, num_of_connections) == -1) {
+        xccl_mhba_error("Failed to listen to server socket errno %d", errno);
+        goto listen_fail;
+    }
+
+    while (num_of_curr_connections < num_of_connections)
+    {
+        /* accept incoming connections */
+        connection[num_of_curr_connections].sock = accept(sock, NULL, 0);
+        connection[num_of_curr_connections].fd = command_fd;
+        connection[num_of_curr_connections].pd_handle = pd_handle;
+        if (connection[num_of_curr_connections].sock != -1)
+        {
+            /* start a new thread but do not wait for it */
+            pthread_create(&thread[num_of_curr_connections], 0, do_sendmsg, (void *)
+            &connection[num_of_curr_connections]);
+            pthread_join(thread[num_of_curr_connections],NULL);
+            num_of_curr_connections += 1;
+        }
+    }
+
+    for(i=0; i<num_of_connections;i++){
+        if (connection[i].return_val != XCCL_OK){
+            xccl_mhba_error("Failed to send cmd_fd");
+            goto listen_fail;
+        }
+    }
+
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close server socket errno %d", errno);
+        if(remove(addr->sun_path)==-1){
+            xccl_mhba_error("Socket file removal failed");
+        }
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    if(remove(addr->sun_path)==-1){
+        xccl_mhba_error("Socket file removal failed");
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    return XCCL_OK;
+
+listen_fail:
+    if(remove(addr->sun_path)==-1){
+        xccl_mhba_error("Socket file removal failed");
+    }
+fail:
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close server socket errno %d", errno);
+    }
+    return XCCL_ERR_NO_MESSAGE;
+}
+
+xccl_status_t xccl_mhba_share_ctx_pd(int root, xccl_mhba_node_t *node, int ctx_fd, uint32_t pd_handle,
+                                     xccl_mhba_context_t *ctx){
+    int      shared_ctx_fd;
+    uint32_t shared_pd_handle;
+    xccl_status_t status;
+    if (root != node->sbgp->group_rank) {
+        sleep(1); //todo change? clients after server
```

Comment:
what is this sleep? if we need to wait for the root, then we should reuse shmem fanin code that we already have for the node group